File: exercicios/aula13/comum/fluxo.py
"""Ponte entre os exercícios da aula 13 e o scanner do projeto.

O scanner **não** é copiado: este módulo importa `src/python/lexer.py` (o lexer
da etapa 1) e oferece por cima dele um fluxo de tokens com lookahead — a
interface `next_token/lookahead` que os exercícios pedem.

Também concentra o que todos os exercícios repetem: leitura do arquivo (ou da
entrada padrão, com `-`), configuração da saída em UTF-8 e o formato dos
diagnósticos, que é o mesmo do projeto:

    Erro léxico na linha L, coluna C: ...
    Erro de sintaxe na linha L, coluna C: esperado X; encontrado Y.
    Erro semântico na linha L, coluna C: ...

Códigos de saída (iguais em todos os exercícios, em Python e em C):
0 = ok, 1 = erro de uso/arquivo, 2 = erro léxico, 3 = erro de sintaxe,
4 = erro semântico.
"""

import sys
from pathlib import Path

RAIZ = Path(__file__).resolve().parents[3]
sys.path.insert(0, str(RAIZ / "src" / "python"))

from lexer import Lexer  # noqa: E402  (depende do ajuste de sys.path acima)

OK, ERRO_USO, ERRO_LEXICO, ERRO_SINTAXE, ERRO_SEMANTICO = 0, 1, 2, 3, 4


def configurar_saida():
    """UTF-8 nas duas saídas, devolvendo intactos bytes inválidos da entrada."""
    sys.stdout.reconfigure(encoding="utf-8", errors="surrogateescape")
    sys.stderr.reconfigure(encoding="utf-8", errors="surrogateescape")


def ler_fonte(caminho: str) -> str:
    """Lê o arquivo (ou a entrada padrão, se `caminho` for "-").

    Lança OSError se o arquivo não puder ser aberto.
    """
    if caminho == "-":
        dados = sys.stdin.buffer.read()
    else:
        dados = Path(caminho).read_bytes()
    # Mesmo tratamento de fim de linha do lexer em C, que lê bytes crus:
    # "\r" é espaço em branco para os dois, então não precisa ser removido.
    return dados.decode("utf-8", errors="surrogateescape")


def tokenizar(fonte: str):
    """Devolve `(tokens, erros_lexicos)`; o último token é sempre EOF."""
    return Lexer(fonte).tokenize()


def descrever(token) -> str:
    """Como um token aparece nas mensagens de erro: `'x'` ou `fim da entrada`."""
    if token.type == "EOF":
        return "fim da entrada"
    return f"'{token.lexeme}'"


def erro_semantico(linha: int, coluna: int, mensagem: str) -> str:
    return f"Erro semântico na linha {linha}, coluna {coluna}: {mensagem}."


class ErroSintaxe(Exception):
    """Interrompe a produção atual; carrega a mensagem já formatada."""

    def __init__(self, token, esperado: str):
        self.token = token
        self.esperado = esperado
        super().__init__(
            f"Erro de sintaxe na linha {token.line}, coluna {token.column}: "
            f"esperado {esperado}; encontrado {descrever(token)}."
        )


class Fluxo:
    """Fluxo de tokens com lookahead (TokenStream)."""

    def __init__(self, tokens):
        self.tokens = tokens
        self.pos = 0

    def atual(self):
        return self.espiar(0)

    def espiar(self, k: int):
        """Lookahead: o k-ésimo token à frente, sem consumir (EOF no fim)."""
        i = self.pos + k
        return self.tokens[i] if i < len(self.tokens) else self.tokens[-1]

    def proximo(self):
        """next_token: consome e devolve o token atual."""
        token = self.atual()
        if token.type != "EOF":
            self.pos += 1
        return token

    def verifica(self, *tipos: str) -> bool:
        return self.atual().type in tipos

    def aceita(self, *tipos: str):
        """Consome o token se for de um dos tipos; senão devolve None."""
        return self.proximo() if self.verifica(*tipos) else None

    def exige(self, tipo: str, esperado: str):
        """Consome um token do tipo pedido ou lança ErroSintaxe."""
        if self.verifica(tipo):
            return self.proximo()
        raise ErroSintaxe(self.atual(), esperado)


def abrir(argv_arquivo: str):
    """Lê e tokeniza; devolve `(Fluxo, codigo)`.

    `codigo` é OK, ou ERRO_USO/ERRO_LEXICO (já com as mensagens impressas em
    stderr) — nesse caso o Fluxo é None e o exercício deve encerrar com ele.
    """
    try:
        fonte = ler_fonte(argv_arquivo)
    except OSError as exc:
        print(f"erro ao abrir '{argv_arquivo}': {exc.strerror}", file=sys.stderr)
        return None, ERRO_USO
    tokens, erros = tokenizar(fonte)
    if erros:
        for erro in erros:
            print(erro, file=sys.stderr)
        return None, ERRO_LEXICO
    return Fluxo(tokens), OK

File: exercicios/aula13/comum/fluxo.c
/* Implementação de fluxo.h — ver o cabeçalho para o contrato. */

#include "fluxo.h"

#include <errno.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void *aloca(size_t n) {
    void *p = malloc(n ? n : 1);
    if (!p) {
        fprintf(stderr, "erro: memória insuficiente\n");
        exit(ERRO_USO);
    }
    return p;
}

void *realoca(void *p, size_t n) {
    void *q = realloc(p, n ? n : 1);
    if (!q) {
        fprintf(stderr, "erro: memória insuficiente\n");
        exit(ERRO_USO);
    }
    return q;
}

char *copia(const char *s) {
    size_t n = strlen(s) + 1;
    char *c = aloca(n);
    memcpy(c, s, n);
    return c;
}

char *formata(const char *fmt, ...) {
    va_list a, b;
    va_start(a, fmt);
    va_copy(b, a);
    int n = vsnprintf(NULL, 0, fmt, b);
    va_end(b);
    char *buf = aloca((size_t)(n < 0 ? 0 : n) + 1);
    vsnprintf(buf, (size_t)n + 1, fmt, a);
    va_end(a);
    return buf;
}

char *ler_fonte(const char *caminho) {
    FILE *arq = strcmp(caminho, "-") == 0 ? stdin : fopen(caminho, "rb");
    if (!arq) {
        fprintf(stderr, "erro ao abrir '%s': %s\n", caminho, strerror(errno));
        return NULL;
    }
    /* Leitura em blocos, que funciona também para stdin (sem fseek). */
    size_t cap = 4096, len = 0;
    char *buf = aloca(cap);
    size_t lidos;
    while ((lidos = fread(buf + len, 1, cap - len - 1, arq)) > 0) {
        len += lidos;
        if (cap - len - 1 == 0) {
            cap *= 2;
            buf = realoca(buf, cap);
        }
    }
    if (arq != stdin) {
        fclose(arq);
    }
    buf[len] = '\0';
    return buf;
}

int fluxo_abrir(Fluxo *f, const char *caminho) {
    f->fonte = ler_fonte(caminho);
    if (!f->fonte) {
        return ERRO_USO;
    }
    lexer_init(&f->lexer, f->fonte);
    lexer_tokenize(&f->lexer);
    f->pos = 0;
    if (f->lexer.errors_len > 0) {
        for (size_t i = 0; i < f->lexer.errors_len; i++) {
            char *s = lexerror_to_string(&f->lexer.errors[i]);
            fprintf(stderr, "%s\n", s);
            free(s);
        }
        fluxo_fechar(f);
        return ERRO_LEXICO;
    }
    return OK;
}

void fluxo_fechar(Fluxo *f) {
    lexer_free(&f->lexer);
    free(f->fonte);
    f->fonte = NULL;
}

const Token *fluxo_espiar(const Fluxo *f, size_t k) {
    size_t i = f->pos + k;
    if (i >= f->lexer.tokens_len) {
        return &f->lexer.tokens[f->lexer.tokens_len - 1]; /* EOF */
    }
    return &f->lexer.tokens[i];
}

const Token *fluxo_atual(const Fluxo *f) { return fluxo_espiar(f, 0); }

const Token *fluxo_proximo(Fluxo *f) {
    const Token *t = fluxo_atual(f);
    if (t->type != TOK_EOF) {
        f->pos++;
    }
    return t;
}

int fluxo_verifica(const Fluxo *f, TokenType tipo) { return fluxo_atual(f)->type == tipo; }

int fluxo_aceita(Fluxo *f, TokenType tipo) {
    if (fluxo_verifica(f, tipo)) {
        fluxo_proximo(f);
        return 1;
    }
    return 0;
}

char *descrever(const Token *t) {
    if (t->type == TOK_EOF) {
        return copia("fim da entrada");
    }
    return formata("'%s'", t->lexeme);
}

void erro_sintaxe(const Token *t, const char *esperado) {
    char *d = descrever(t);
    fprintf(stderr, "Erro de sintaxe na linha %d, coluna %d: esperado %s; encontrado %s.\n",
            t->line, t->column, esperado, d);
    free(d);
}
